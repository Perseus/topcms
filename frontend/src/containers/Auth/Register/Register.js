import { mapState } from 'vuex';
import errorHandlerMixin from '../../../mixins/errorHandler';

const Register = {
  created() {
    this.$store.watch((state) => state.user.authenticationStatus, (oldVal, newVal) => {
      console.log(oldVal, newVal);
    });
  },
  data() {
    return {
      isLoading: false,
      username: '',
      email: '',
      emailRepeat: '',
      password: '',
    };
  },
  mixins: [ errorHandlerMixin ],

  methods: {
    registerUser() {

      if (this.email !== this.emailRepeat) {
        this.setError('emailRepeat', 'The repeated email does not match the original email.');
      } else {

        const userDetails = {
          username: this.username,
          password: this.password,
          email: this.email
        };

        this.$store.dispatch('registerUser', userDetails);
      }
    },

  },
  computed: {
    errorComputer() {
      console.log(this.user);
    },
    ...mapState({
      'user': state => state.user,
      'isLoggingIn': state => state.user.authenticationStatus.isLoggingIn
    }),
  },

  watch: {
    user() {
      console.log('USER CHANGED');
    }
  }
};

export default Register;