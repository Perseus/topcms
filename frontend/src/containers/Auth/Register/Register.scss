.register {
  align-items: center;
  display: flex;
  flex: 1;
  justify-content: center;
  flex-direction: column;

  &__form {
    background-color: $color-white;
    padding: 4rem;
    box-shadow: 0px 2px 4px rgba($color-primary-1, .5);
    border-radius: 4px;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;

  }
  
  &__title {
    font-size: 2rem;
    margin-bottom: 1.2rem;
    color: $color-grey-1;
    font-weight: 400;
    text-transform: uppercase;
  }

  &__subtitle {
    color: $color-grey-6;
    font-size: 1.2rem;
    font-weight: 400;
    padding-bottom: 4rem;
  }

  &__field {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
  }

  &__field-label {
    font-size: 1.2rem;
    font-weight: 300;
    color: $color-grey-6;
    margin-bottom: .7rem;
  }

  &__field-input-text {
    border: 1px solid rgba( $color-grey-8, .25);
    font-size: 1.2rem;
    width: 32rem;
    padding: 1rem 0px;
    margin-bottom: 2rem;
    color: $color-grey-4;
    padding-left: .5rem;
  }

  &__submit {
    background-color: $color-primary-6;
    width: 32rem;
    height: 5rem;
    border: none;
    border-radius: 8px;
    color: $color-white;
    text-transform: uppercase;
    font-weight: bold;
    font-size: 1.4rem;
    display: flex;
    justify-content: center;
    align-items: center;
    font-weight: 500;
    
    &:hover {
      cursor: pointer;
    }

  }

  &__login-link {
    margin-top: 1.4rem;
    font-size: 1.3rem;
    color: $color-primary-6;
    font-weight: 500;

    &:hover {
      cursor: pointer;
    }
  }



}