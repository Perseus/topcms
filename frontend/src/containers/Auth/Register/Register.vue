<template>
  <div class="register">
    

    <form @submit.prevent="registerUser" class="register__form">
      <h1 class="register__title">Register</h1>
      <h2 class="register__subtitle">Please enter your registration details</h2>
      <div class="register__field">
        <label for="register__username-input" class="register__field-label">Username</label>
        <input type="text" id="register__username-input" class="register__field-input-text" />
      </div>
      <div class="register__field">
        <label for="register__username-input" class="register__field-label">eMail</label>
        <input type="email" id="register__username-input" class="register__field-input-text" />
      </div>
      <div class="register__field">
        <label for="register__username-input" class="register__field-label">Repeat eMail</label>
        <input type="email" id="register__username-input" class="register__field-input-text" />
      </div>
      <div class="register__field">
        <label for="register__username-input" class="register__field-label">Password</label>
        <input type="password" id="register__username-input" class="register__field-input-text" />
      </div>
      <button class="register__submit" type="submit">
        <span v-if="!isLoading">Register</span>
        <div class="loading-circle" v-else>

        </div>
        </button>
    </form>
  </div>
</template>


<script src="./Register.js"></script>
<style src="./Register.scss" lang="scss"></style>