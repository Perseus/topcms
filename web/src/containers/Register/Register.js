import { mapActions, mapGetters } from 'vuex';

import ActionTypes from '../../store/types/ActionTypes';
import { changeRoute } from '../../utils/RouterUtils';
import RouteNames from '../../config/RouteNames';
import TInput from '../../components/ValidationInputs/TInput.vue';

const Register = {

  name: 'register-form',
  data() {
    return {
      username: '',
      password: '',
      email: '',
    };
  },
  components: {
    TInput
  },
  methods: {
    ...getActionDispatchers(),

    redirectToLogin() {
      this.changeRoute( {
        name: RouteNames.AUTH.LOGIN
      } );
    },

    onUserRegister() {
      this.registerUser( {
        username: this.username,
        password: this.password,
        email: this.email
      } );
    }
  },

  computed: {
    ...getStateGetters(),
  },

};

function getActionDispatchers() {
  return mapActions( {
    registerUser: ActionTypes.registerUser,
    changeRoute: ActionTypes.changeRoute,
  } );
}

function getStateGetters() {
  return mapGetters( {
    isUserRegistering: 'isUserRegistering'
  } );
}

export default Register;
