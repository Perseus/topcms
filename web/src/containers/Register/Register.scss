.register-form {
  display: flex;
  justify-content: center;
  align-items: center;
}

.cta {
  display: flex;
  justify-content: center;
  flex-direction: column;
  align-items: center;
}

.have-account-cta {
  margin: 10px 0;
}