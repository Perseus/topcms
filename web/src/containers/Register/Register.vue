<template>
  <section class="container register-form">
    <div class="card">
      <div class="card-content">
        <form @submit.prevent="onUserRegister">
          <b-field
            label="Username"
            :type="{ 'is-danger': errors.has('username') }"
            :message="errors.first('username')"
          >
            <b-input
              v-validate="'required'"
              data-test="username"
              name="username"
              v-model="username"
            ></b-input>
          </b-field>
          <b-field
            label="Email"
            :type="{ 'is-danger': errors.has('email') }"
            :message="errors.first('email')"
          >
            <b-input v-validate="'required|email'" name="email" v-model="email"></b-input>
          </b-field>
          <b-field
            label="Password"
            :type="{ 'is-danger': errors.has('password') }"
            :message="errors.first('password')"
          >
            <b-input
              v-model="password"
              v-validate="'required|min:8'"
              name="password"
              type="password"
              password-reveal
            ></b-input>
          </b-field>
          <div class="cta">
            <a
              href="#"
              class="have-account-cta"
              @click.prevent="redirectToLogin"
            >Already have an account ?</a>
            <b-button type="is-primary" native-type="submit" :loading="isUserRegistering">Register</b-button>
          </div>
        </form>
      </div>
    </div>
  </section>
</template>


<script src="./Register.js"></script>
<style src="./Register.scss" lang="scss" scoped></style>