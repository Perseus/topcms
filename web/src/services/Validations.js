import { extend } from 'vee-validate';
import {
  required, min, email, confirmed, required_if, numeric, integer
} from 'vee-validate/dist/rules';


extend( 'required', {
  ...required,

  message: fieldName => `${fieldName} is required`
} );

extend( 'min', {
  ...min,
  message: ( fieldName, args ) => `${fieldName} needs to be at least ${args.length} characters long`
} );

extend( 'email', {
  ...email,
  message: fieldName => `${fieldName} needs to be a valid e-mail`,
} );

extend( 'confirmed', {
  ...confirmed,
  message: ( fieldName, args ) => `This field's value needs to match the value in ${args.target}`
} );

extend( 'required_if', required_if );

extend( 'numeric', {
  ...numeric,
  message: fieldName => `${fieldName} needs to be a valid number`
} );

extend( 'integer', {
  ...integer,
  message: fieldName => `${fieldName} needs to be a valid number`
} );

extend( 'is_url', {
  validate: ( value ) => {
    let url;
    try {
      url = new URL( value );
    } catch ( _ ) {
      return false;
    }

    return url.protocol === 'http:' || url.protocol === 'https:';
  },
  message: fieldName => `${fieldName} needs to be a valid URL`
} );
