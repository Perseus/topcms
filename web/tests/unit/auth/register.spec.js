import { mount, createLocalVue } from '@vue/test-utils';
import { ValidationObserver } from 'vee-validate';
//import Buefy from 'buefy';

import Vuex from 'vuex';
import faker from 'faker';

import '@services/Validations';

import Register from '@containers/Register/Register.vue';

const flushPromises = () => new Promise( setTimeout );

const localVue = createLocalVue();
localVue.use( Vuex );
localVue.use( Buefy );

localVue.component( 'ValidationObserver', ValidationObserver );

const registerUser = jest.fn();

function mountComponentWithLocalVue( component, options ) {
  const store = new Vuex.Store( {
    getters: {
      isUserRegistering: () => false,
    },
    actions: {
      registerUser
    }
  } );

  return mount( component, {
    localVue,
    store,
    ...options,
  } );
}

describe( 'Register Component', () => {
  it( 'renders correctly', () => {
    const wrapper = mountComponentWithLocalVue( Register );

    expect( wrapper ).toMatchSnapshot();
  } );

  it( 'has validation errors if all fields are empty', async() => {
    const wrapper = mountComponentWithLocalVue( Register );

    const submitButton = wrapper.find( '[data-test="register-button"]' );
    await submitButton.trigger( 'submit' );

    await flushPromises();

    const errorElements = wrapper.findAll( '.help' );
    errorElements.wrappers.forEach( wrap => expect( wrap.text() ).not.toHaveLength( 0 ) );
  } );

  it( 'has validation errors if password doesn\'t match confirm password', async() => {
    const wrapper = mountComponentWithLocalVue( Register, {
      data: () => ( {
        username: faker.name.findName(),
        email: faker.internet.email(),
        password: faker.internet.password(),
      } )
    } );

    await wrapper.vm.$nextTick();

    const confirmPasswordField = wrapper.find( '[data-test="Confirm Password"]' );
    await confirmPasswordField.setValue( 'foobar' );

    const submitButton = wrapper.find( '[data-test="register-button"]' );
    await submitButton.trigger( 'submit' );

    await flushPromises();

    const confirmPasswordErrorEl = wrapper.find( '.help.is-danger' );
    expect( confirmPasswordErrorEl.text() ).toContain( 'match the value' );
  } );

  it( 'triggers registerUser action on successful form submit', async() => {
    const data = {
      username: faker.name.findName(),
      email: faker.internet.email(),
      password: faker.internet.password(),
    };

    const wrapper = mountComponentWithLocalVue( Register, {
      data: () => data,
    } );

    await wrapper.vm.$nextTick();

    const confirmPasswordField = wrapper.find( '[data-test="Confirm Password"]' );
    await confirmPasswordField.setValue( data.password );

    const submitButton = wrapper.find( '[data-test="register-button"]' );
    await submitButton.trigger( 'submit' );

    await flushPromises();

    expect( registerUser ).toHaveBeenCalled();
  } );
} );
