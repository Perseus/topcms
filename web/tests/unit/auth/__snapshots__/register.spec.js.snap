// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Component renders correctly 1`] = `
<section class="container register-form">
  <div class="card">
    <div class="card-content"><span><form><span label="Username" name="Username"><div class="field" name="Username"><label class="label">Username</label> <div class="control"><input type="text" autocomplete="off" data-test="Username" label="Username" name="Username" class="input"> <!----> <!----> <!----></div> <p class="help"></p></div></span> <span label="E-mail" name="E-mail"><div class="field" name="E-mail"><label class="label">E-mail</label> <div class="control"><input type="text" autocomplete="off" data-test="E-mail" label="E-mail" name="E-mail" class="input"> <!----> <!----> <!----></div> <p class="help"></p></div></span> <span label="Password" type="password" name="Password"><div class="field" name="Password"><label class="label">Password</label> <div class="control"><input type="password" autocomplete="off" data-test="Password" label="Password" name="Password" class="input"> <!----> <!----> <!----></div> <p class="help"></p></div></span> <span label="Confirm Password" type="password" name="Confirm Password"><div class="field" name="Confirm Password"><label class="label">Confirm Password</label> <div class="control"><input type="password" autocomplete="off" data-test="Confirm Password" label="Confirm Password" name="Confirm Password" class="input"> <!----> <!----> <!----></div> <p class="help"></p></div></span>
      <div class="cta"><a href="#" class="have-account-cta">Already have an account ?</a> <button type="submit" data-test="register-button" class="button is-primary">
          <!----> <span>Register</span>
          <!----></button></div>
      </form>
      </span>
    </div>
  </div>
</section>
`;
